File: CIA - 1/6_oracle.py
import networkx as nx
import matplotlib.pyplot as plt

# Oracle algorithm (direct knowledge of the best path)
def oracle_search(graph, start, goal, predefined_path):
    if predefined_path[0] == start and predefined_path[-1] == goal:
        return predefined_path
    else:
        print("Predefined path doesn't match the search goal.")
        return None

# Input graph details and predefined path
num_vertices = int(input("Enter the number of vertices: "))
edges = []
for _ in range(int(input("Enter the number of edges: "))):
    edge = input("Enter edge (format: source target weight): ").split()
    edges.append((edge[0], edge[1], int(edge[2])))

start_vertex = input("Enter the start vertex: ")
goal_vertex = input("Enter the goal vertex: ")
predefined_path = input("Enter predefined path as space-separated vertices: ").split()

# Create a weighted graph using networkx
G = nx.Graph()
for edge in edges:
    G.add_edge(edge[0], edge[1], weight=edge[2])

# Perform Oracle Search
path = oracle_search(G, start_vertex, goal_vertex, predefined_path)

# Draw the initial graph
pos = nx.spring_layout(G)
plt.figure(figsize=(10, 7))
nx.draw(G, pos, with_labels=True, node_size=700, node_color='lightblue', font_size=16, font_weight='bold', edge_color='gray')
edge_labels = nx.get_edge_attributes(G, 'weight')
nx.draw_networkx_edge_labels(G, pos, edge_labels=edge_labels)
plt.title("Initial Graph")
plt.show()

# If a path is found, display the path
if path:
    path_edges = list(zip(path, path[1:]))
    plt.figure(figsize=(10, 7))
    nx.draw(G, pos, with_labels=True, node_size=700, node_color='lightblue', font_size=16, font_weight='bold', edge_color='gray')
    nx.draw_networkx_edges(G, pos, edgelist=path_edges, edge_color='r', width=2)
    nx.draw_networkx_nodes(G, pos, nodelist=path, node_color='lightgreen')
    plt.title(f"Oracle Path: {path}")
    plt.show()
else:
    print("No path found.")

File: CIA - 1/6_oracle.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX 100

int adjMatrix[MAX][MAX], numVertices;
char vertices[MAX];

// Simulated Oracle path (predefined by the oracle)
int oraclePath[MAX], pathLength = 0;

// Function to display the graph as adjacency matrix
void displayGraph() {
    printf("\nAdjacency Matrix:\n");
    for (int i = 0; i < numVertices; i++) {
        printf("%c: ", vertices[i]);
        for (int j = 0; j < numVertices; j++) {
            printf("%d ", adjMatrix[i][j]);
        }
        printf("\n");
    }
}

// Oracle search function (simulated)
void oracleSearch(int source, int goal) {
    printf("Oracle knows the path from %c to %c:\n", vertices[source], vertices[goal]);
    for (int i = 0; i < pathLength; i++) {
        printf("%c ", vertices[oraclePath[i]]);
        if (i < pathLength - 1) {
            printf("-> ");
        }
    }
    printf("\nGoal %c reached by Oracle!\n", vertices[goal]);
}

// Main function
int main() {
    int i, j, source, destination;
    char start, goal;

    printf("Enter the number of vertices: ");
    scanf("%d", &numVertices);

    printf("Enter the names of the vertices:\n");
    for (i = 0; i < numVertices; i++) {
        scanf(" %c", &vertices[i]);
    }

    printf("Enter the number of edges: ");
    int numEdges;
    scanf("%d", &numEdges);

    // Initialize adjacency matrix
    for (i = 0; i < numVertices; i++) {
        for (j = 0; j < numVertices; j++) {
            adjMatrix[i][j] = 0;
        }
    }

    printf("Enter the edges (source destination weight):\n");
    for (i = 0; i < numEdges; i++) {
        char s, d;
        int weight;
        scanf(" %c %c %d", &s, &d, &weight);
        int source = strchr(vertices, s) - vertices;
        int destination = strchr(vertices, d) - vertices;
        adjMatrix[source][destination] = weight;
        adjMatrix[destination][source] = weight; // Assuming undirected graph
    }

    displayGraph();

    // Taking start and goal inputs
    printf("Enter the start vertex: ");
    scanf(" %c", &start);
    printf("Enter the goal vertex: ");
    scanf(" %c", &goal);

    source = strchr(vertices, start) - vertices;
    destination = strchr(vertices, goal) - vertices;

    // Simulate an Oracle knowing the path (predefined for simplicity)
    printf("Enter the length of the Oracle's known path: ");
    scanf("%d", &pathLength);
    printf("Enter the Oracle's known path (sequence of vertices):\n");
    for (i = 0; i < pathLength; i++) {
        char vertex;
        scanf(" %c", &vertex);
        oraclePath[i] = strchr(vertices, vertex) - vertices;
    }

    // Perform Oracle Search
    oracleSearch(source, destination);

    return 0;
}
